File: src/utils/data analysis/plotter_mpi.R
# =============================================================================
# MPI Count Sketch - Professional R Plotter (Enhanced)
# =============================================================================
# 
# Usage: Run in R terminal from project root:
#   source("src/utils/data analysis/plotter_mpi.R")
#
# Required packages (install first time):
#   install.packages(c("ggplot2", "dplyr", "tidyr", "scales"))
#
# Output: charts/*.png
# =============================================================================

# --- Setup -------------------------------------------------------------------
library(ggplot2)
library(dplyr)
library(scales)

# Create charts directory
dir.create("charts", showWarnings = FALSE)

# Publication theme with enhanced styling
theme_publication <- function() {
  theme_minimal(base_size = 12) +
  theme(
    plot.title = element_text(face = "bold", size = 14, hjust = 0.5),
    plot.subtitle = element_text(face = "italic", size = 10, hjust = 0.5, color = "gray40"),
    axis.title = element_text(face = "bold", size = 11),
    axis.text = element_text(size = 10),
    legend.position = "top",
    legend.title = element_blank(),
    legend.text = element_text(size = 10),
    panel.grid.minor = element_blank(),
    panel.grid.major = element_line(color = "gray90"),
    panel.border = element_rect(color = "grey70", fill = NA, linewidth = 0.5),
    plot.margin = margin(10, 15, 10, 10)
  )
}

# Color palette
col_compute <- "#2E86AB"
col_comm <- "#3CB371"
col_ideal <- "#E63946"
colors_size <- c("500K" = "#2E86AB", "1M" = "#F18F01", "8M" = "#A23B72")

# =============================================================================
# 1. STRONG SCALING PLOTS
# =============================================================================

# Load 8M strong scaling data
perf_8m <- read.csv("results/performance_mpi_8388608.csv")

# --- Plot 1: Strong Scaling Speedup (Compute Only) ---------------------------
p1 <- ggplot(perf_8m, aes(x = Processes)) +
  geom_line(aes(y = Speedup_Compute, color = "MPI Speedup", linetype = "MPI Speedup"), linewidth = 1.2) +
  geom_point(aes(y = Speedup_Compute, color = "MPI Speedup"), size = 3) +
  geom_line(aes(y = Processes, color = "Ideal Linear", linetype = "Ideal Linear"), linewidth = 1) +
  scale_x_continuous(breaks = perf_8m$Processes, trans = "log2") +
  scale_y_continuous(breaks = seq(0, 70, 10)) +
  scale_color_manual(values = c("MPI Speedup" = col_compute, "Ideal Linear" = col_ideal)) +
  scale_linetype_manual(values = c("MPI Speedup" = "solid", "Ideal Linear" = "dashed")) +
  labs(
    title = "Strong Scaling: Speedup (Compute Only)",
    subtitle = expression(paste("8M items | Formula: ", S[p], " = ", T[1], " / ", T[p])),
    x = "Number of MPI Processes",
    y = "Speedup Factor"
  ) +
  theme_publication()

ggsave("charts/strong_scaling_speedup_compute.png", p1, width = 10, height = 6, dpi = 300)
print("Saved: strong_scaling_speedup_compute.png")

# --- Plot 2: Strong Scaling Speedup (With Communication) ---------------------
p2 <- ggplot(perf_8m, aes(x = Processes)) +
  geom_line(aes(y = Speedup_Comm, color = "MPI Speedup", linetype = "MPI Speedup"), linewidth = 1.2) +
  geom_point(aes(y = Speedup_Comm, color = "MPI Speedup"), size = 3) +
  geom_line(aes(y = Processes, color = "Ideal Linear", linetype = "Ideal Linear"), linewidth = 1) +
  scale_x_continuous(breaks = perf_8m$Processes, trans = "log2") +
  scale_y_continuous(breaks = seq(0, 70, 10)) +
  scale_color_manual(values = c("MPI Speedup" = col_comm, "Ideal Linear" = col_ideal)) +
  scale_linetype_manual(values = c("MPI Speedup" = "solid", "Ideal Linear" = "dashed")) +
  labs(
    title = "Strong Scaling: Speedup (With Communication)",
    subtitle = "8M items | Includes MPI communication overhead",
    x = "Number of MPI Processes",
    y = "Speedup Factor"
  ) +
  theme_publication()

ggsave("charts/strong_scaling_speedup.png", p2, width = 10, height = 6, dpi = 300)
print("Saved: strong_scaling_speedup.png")

# --- Plot 3: Strong Scaling Efficiency ----------------------------------------
p3 <- ggplot(perf_8m, aes(x = Processes)) +
  geom_line(aes(y = Efficiency_Compute, color = "Compute Only"), linewidth = 1.2) +
  geom_point(aes(y = Efficiency_Compute, color = "Compute Only"), size = 3) +
  geom_line(aes(y = Efficiency_Comm, color = "With Communication"), linewidth = 1.2) +
  geom_point(aes(y = Efficiency_Comm, color = "With Communication"), size = 3) +
  geom_hline(yintercept = 100, linetype = "dashed", color = col_ideal, linewidth = 1) +
  scale_x_continuous(breaks = perf_8m$Processes, trans = "log2") +
  scale_y_continuous(limits = c(0, 120)) +
  scale_color_manual(values = c("Compute Only" = col_compute, "With Communication" = col_comm)) +
  labs(
    title = "Strong Scaling: Efficiency",
    subtitle = expression(paste("8M items | Formula: ", E[p], " = ", S[p], " / p × 100% | Goal: Stay close to 100%")),
    x = "Number of MPI Processes",
    y = "Efficiency (%)"
  ) +
  annotate("text", x = 48, y = 104, label = "Ideal (100%)", color = col_ideal, fontface = "italic", size = 3.5) +
  theme_publication()

ggsave("charts/strong_scaling_efficiency.png", p3, width = 10, height = 6, dpi = 300)
print("Saved: strong_scaling_efficiency.png")

# --- Plot 4: Strong Scaling Time ---------------------------------------------
p4 <- ggplot(perf_8m, aes(x = Processes)) +
  geom_line(aes(y = Compute_Time, color = "Compute Only"), linewidth = 1.2) +
  geom_point(aes(y = Compute_Time, color = "Compute Only"), size = 3) +
  geom_line(aes(y = Total_Time, color = "Total (with comm)"), linewidth = 1.2) +
  geom_point(aes(y = Total_Time, color = "Total (with comm)"), size = 3) +
  scale_x_continuous(breaks = perf_8m$Processes, trans = "log2") +
  scale_y_log10(labels = scales::number_format(accuracy = 0.01)) +
  scale_color_manual(values = c("Compute Only" = col_compute, "Total (with comm)" = col_comm)) +
  labs(
    title = "Strong Scaling: Execution Time",
    subtitle = "8M items | Communication overhead gap increases with process count",
    x = "Number of MPI Processes",
    y = "Time (seconds, log scale)"
  ) +
  theme_publication()

ggsave("charts/strong_scaling_time.png", p4, width = 10, height = 6, dpi = 300)
print("Saved: strong_scaling_time.png")

# =============================================================================
# 2. WEAK SCALING PLOTS
# =============================================================================

# Load 8M weak scaling data
weak_8m <- read.csv("results/weak_scaling_mpi_8388608.csv")

# Calculate CORRECT weak scaling efficiency: T1_base / Tp
weak_8m <- weak_8m %>%
  mutate(
    T1_base = Serial_Time / Processes,
    Efficiency_Compute = T1_base / Compute_Time,
    Efficiency_Comm = T1_base / Compute_Communication_Time
  )

# --- Plot 5: Weak Scaling Efficiency (Compute Only) --------------------------
p5 <- ggplot(weak_8m, aes(x = Processes)) +
  geom_line(aes(y = Efficiency_Compute, color = "Measured Efficiency"), linewidth = 1.2) +
  geom_point(aes(y = Efficiency_Compute, color = "Measured Efficiency"), size = 3) +
  geom_hline(aes(yintercept = 1, color = "Ideal Scaling (1.0)"), linetype = "dashed", linewidth = 1) +
  scale_x_continuous(breaks = weak_8m$Processes, trans = "log2") +
  scale_y_continuous(limits = c(0, 1.4)) +
  scale_color_manual(values = c("Measured Efficiency" = col_compute, "Ideal Scaling (1.0)" = col_ideal)) +
  labs(
    title = "Weak Scaling: Efficiency (Compute Only)",
    subtitle = expression(paste("Formula: ", S[w], " = ", T[1](N), " / ", T[p](N %.% p), " | Goal: Stay close to 1.0")),
    x = "Number of MPI Processes",
    y = expression(paste("Efficiency (", T[1], " / ", T[p], ")"))
  ) +
  theme_publication()

ggsave("charts/weak_scaling_efficiency_compute.png", p5, width = 10, height = 6, dpi = 300)
print("Saved: weak_scaling_efficiency_compute.png")

# --- Plot 6: Weak Scaling Efficiency (With Communication) --------------------
p6 <- ggplot(weak_8m, aes(x = Processes)) +
  geom_line(aes(y = Efficiency_Comm, color = "Measured Weak Efficiency"), linewidth = 1.2) +
  geom_point(aes(y = Efficiency_Comm, color = "Measured Weak Efficiency"), size = 3) +
  geom_hline(aes(yintercept = 1, color = "Ideal Scaling (1.0)"), linetype = "dashed", linewidth = 1) +
  scale_x_continuous(breaks = weak_8m$Processes, trans = "log2") +
  scale_y_continuous(limits = c(0, 1.4)) +
  scale_color_manual(values = c("Measured Weak Efficiency" = col_comm, "Ideal Scaling (1.0)" = col_ideal)) +
  labs(
    title = "Weak Scaling: Efficiency vs. Processes",
    subtitle = "Includes communication overhead | Goal: Stay close to 1.0",
    x = "Number of MPI Processes",
    y = expression(paste("Efficiency (", T[1], " / ", T[p], ")"))
  ) +
  theme_publication()

ggsave("charts/weak_scaling_efficiency.png", p6, width = 10, height = 6, dpi = 300)
print("Saved: weak_scaling_efficiency.png")

# --- Plot 7: Weak Scaling Times ---------------------------------------------
p7 <- ggplot(weak_8m, aes(x = Processes)) +
  geom_line(aes(y = Compute_Time, color = "Compute Only"), linewidth = 1.2) +
  geom_point(aes(y = Compute_Time, color = "Compute Only"), size = 3) +
  geom_line(aes(y = Compute_Communication_Time, color = "Total (with comm)"), linewidth = 1.2) +
  geom_point(aes(y = Compute_Communication_Time, color = "Total (with comm)"), size = 3) +
  scale_x_continuous(breaks = weak_8m$Processes, trans = "log2") +
  scale_color_manual(values = c("Compute Only" = col_compute, "Total (with comm)" = col_comm)) +
  labs(
    title = "Weak Scaling: Execution Times",
    subtitle = "Ideal: constant time | Reality: communication overhead grows with P",
    x = "Number of MPI Processes",
    y = "Time (seconds)"
  ) +
  theme_publication()

ggsave("charts/weak_scaling_times.png", p7, width = 10, height = 6, dpi = 300)
print("Saved: weak_scaling_times.png")

# =============================================================================
# 3. COMPARISON PLOTS (Multiple Dataset Sizes)
# =============================================================================

# Load all strong scaling data
perf_500k <- read.csv("results/performance_mpi_524288.csv") %>% mutate(Size = "500K")
perf_1m <- read.csv("results/performance_mpi_1048576.csv") %>% mutate(Size = "1M")
perf_8m_comp <- perf_8m %>% mutate(Size = "8M")

perf_all <- bind_rows(perf_500k, perf_1m, perf_8m_comp)
perf_all$Size <- factor(perf_all$Size, levels = c("500K", "1M", "8M"))

# --- Plot 8: Strong Scaling Comparison (3 sizes) -----------------------------
p8 <- ggplot(perf_all, aes(x = Processes, y = Speedup_Comm, color = Size)) +
  geom_line(linewidth = 1.2) +
  geom_point(size = 3) +
  geom_line(aes(y = Processes), linetype = "dashed", color = col_ideal, linewidth = 1, show.legend = FALSE) +
  scale_x_continuous(breaks = c(2, 4, 8, 16, 32, 64), trans = "log2") +
  scale_color_manual(values = colors_size) +
  labs(
    title = "Strong Scaling Comparison: Speedup vs. Dataset Size",
    subtitle = "Larger datasets achieve better scaling | 500K shows degradation at 64 processes",
    x = "Number of MPI Processes",
    y = "Speedup Factor (with communication)",
    color = "Dataset Size"
  ) +
  theme_publication() +
  theme(legend.title = element_text(face = "bold"))

ggsave("charts/strong_comparison_mpi.png", p8, width = 10, height = 6, dpi = 300)
print("Saved: strong_comparison_mpi.png")

# --- Plot 9: Strong Scaling Efficiency Comparison ----------------------------
p9_eff <- ggplot(perf_all, aes(x = Processes, y = Efficiency_Comm, color = Size)) +
  geom_line(linewidth = 1.2) +
  geom_point(size = 3) +
  geom_hline(yintercept = 100, linetype = "dashed", color = col_ideal, linewidth = 1) +
  scale_x_continuous(breaks = c(2, 4, 8, 16, 32, 64), trans = "log2") +
  scale_color_manual(values = colors_size) +
  labs(
    title = "Strong Scaling: Efficiency Comparison",
    subtitle = "Communication overhead impact varies with problem size",
    x = "Number of MPI Processes",
    y = "Efficiency (%)",
    color = "Dataset Size"
  ) +
  annotate("text", x = 48, y = 105, label = "Ideal (100%)", color = col_ideal, fontface = "italic", size = 3.5) +
  theme_publication() +
  theme(legend.title = element_text(face = "bold"))

ggsave("charts/strong_efficiency_comparison.png", p9_eff, width = 10, height = 6, dpi = 300)
print("Saved: strong_efficiency_comparison.png")

# --- Plot 10: Weak Scaling Comparison (3 sizes) ------------------------------
weak_500k <- read.csv("results/weak_scaling_mpi_524288.csv") %>% mutate(Size = "500K")
weak_1m <- read.csv("results/weak_scaling_mpi_1048576.csv") %>% mutate(Size = "1M")
weak_8m_comp <- weak_8m %>% mutate(Size = "8M")

weak_all <- bind_rows(weak_500k, weak_1m, weak_8m_comp)
weak_all$Size <- factor(weak_all$Size, levels = c("500K", "1M", "8M"))

# Calculate efficiency for all sizes
weak_all <- weak_all %>%
  mutate(
    T1_base = Serial_Time / Processes,
    Efficiency = T1_base / Compute_Communication_Time
  )

p10 <- ggplot(weak_all, aes(x = Processes, y = Efficiency, color = Size)) +
  geom_line(linewidth = 1.2) +
  geom_point(size = 3) +
  geom_hline(yintercept = 1, linetype = "dashed", color = col_ideal, linewidth = 1) +
  scale_x_continuous(breaks = c(2, 4, 8, 16, 32, 64), trans = "log2") +
  scale_y_continuous(limits = c(0, 1.4)) +
  scale_color_manual(values = colors_size) +
  labs(
    title = "Weak Scaling Comparison: Efficiency vs. Dataset Size",
    subtitle = expression(paste("Formula: ", S[w], " = ", T[1](N), " / ", T[p](N %.% p), " | Goal: Stay close to 1.0")),
    x = "Number of MPI Processes",
    y = expression(paste("Efficiency (", T[1], " / ", T[p], ")")),
    color = "Base Dataset"
  ) +
  annotate("text", x = 48, y = 1.05, label = "Ideal (1.0)", color = col_ideal, fontface = "italic", size = 3.5) +
  theme_publication() +
  theme(legend.title = element_text(face = "bold"))

ggsave("charts/weak_comparison_mpi.png", p10, width = 10, height = 6, dpi = 300)
print("Saved: weak_comparison_mpi.png")

# =============================================================================
# DONE
# =============================================================================
cat("\n========================================\n")
cat("All 10 charts generated successfully!\n")
cat("Output folder: charts/\n")
cat("========================================\n")
cat("\nCharts saved:\n")
cat("  1. strong_scaling_speedup_compute.png\n")
cat("  2. strong_scaling_speedup.png\n")
cat("  3. strong_scaling_efficiency.png\n")
cat("  4. strong_scaling_time.png\n")
cat("  5. weak_scaling_efficiency_compute.png\n")
cat("  6. weak_scaling_efficiency.png\n")
cat("  7. weak_scaling_times.png\n")
cat("  8. strong_comparison_mpi.png\n")
cat("  9. strong_efficiency_comparison.png\n")
cat(" 10. weak_comparison_mpi.png\n")

File: src/utils/data analysis/plotter_mpi.py
#!/usr/bin/env python3
"""
MPI Count Sketch - Plotter Script

Generates charts for MPI benchmarks:
  1. strong_scaling_speedup_compute.png
  2. strong_scaling_speedup.png (with communication)
  3. strong_scaling_time.png
  4. weak_scaling_efficiency_compute.png
  5. weak_scaling_efficiency.png (with communication)
  6. weak_scaling_times.png
  7. strong_comparison_mpi.png (3 sizes comparison)
  8. weak_comparison_mpi.png (optional)

Usage:
    python src/utils/data\ analysis/plotter_mpi.py
"""

import pandas as pd
import matplotlib.pyplot as plt
import os
import re

# Output directory for charts
CHARTS_DIR = "charts"
os.makedirs(CHARTS_DIR, exist_ok=True)


def extract_lines_from_filename(csv_file):
    """Extract number of lines from filename like performance_mpi_8388608.csv"""
    match = re.search(r"_(\d+)\.csv$", os.path.basename(csv_file))
    return int(match.group(1)) if match else None


def generate_strong_scaling_plots(csv_file):
    """
    Generate Strong Scaling charts for MPI.
    For pure MPI, we only have one data point per process count (N_Threads=1).
    """
    if not os.path.exists(csv_file):
        print(f"Warning: {csv_file} not found.")
        return
    
    df = pd.read_csv(csv_file)
    df = df.sort_values('Processes')
    total_lines = extract_lines_from_filename(csv_file)
    
    print(f"Generating strong scaling plots for {total_lines} items...")
    
    # 1. Strong Scaling Speedup (Compute Only)
    plt.figure(figsize=(10, 6))
    plt.plot(df['Processes'], df['Speedup_Compute'], 
             marker='o', linewidth=2, markersize=8, color='blue', label='MPI Speedup')
    
    # Ideal line
    max_p = df['Processes'].max()
    min_p = df['Processes'].min()
    plt.plot([min_p, max_p], [min_p, max_p], 
             'r--', linewidth=1.5, alpha=0.7, label='Ideal Linear Speedup')
    
    plt.title(f'Strong Scaling: Speedup (Compute Only) - {total_lines} items', fontsize=14)
    plt.xlabel('Number of MPI Processes', fontsize=12)
    plt.ylabel('Speedup Factor', fontsize=12)
    plt.xscale('log', base=2)
    plt.xticks(df['Processes'], df['Processes'])
    plt.legend()
    plt.grid(True, which="both", ls="-", alpha=0.5)
    plt.tight_layout()
    plt.savefig(f'{CHARTS_DIR}/strong_scaling_speedup_compute.png', dpi=300)
    print(f"  Saved: strong_scaling_speedup_compute.png")
    plt.close()
    
    # 2. Strong Scaling Speedup (With Communication)
    plt.figure(figsize=(10, 6))
    plt.plot(df['Processes'], df['Speedup_Comm'], 
             marker='s', linewidth=2, markersize=8, color='green', label='MPI Speedup (with comm)')
    plt.plot([min_p, max_p], [min_p, max_p], 
             'r--', linewidth=1.5, alpha=0.7, label='Ideal Linear Speedup')
    
    plt.title(f'Strong Scaling: Speedup (With Communication) - {total_lines} items', fontsize=14)
    plt.xlabel('Number of MPI Processes', fontsize=12)
    plt.ylabel('Speedup Factor', fontsize=12)
    plt.xscale('log', base=2)
    plt.xticks(df['Processes'], df['Processes'])
    plt.legend()
    plt.grid(True, which="both", ls="-", alpha=0.5)
    plt.tight_layout()
    plt.savefig(f'{CHARTS_DIR}/strong_scaling_speedup.png', dpi=300)
    print(f"  Saved: strong_scaling_speedup.png")
    plt.close()
    
    # 3. Strong Scaling Execution Time
    plt.figure(figsize=(10, 6))
    plt.plot(df['Processes'], df['Compute_Time'], 
             marker='o', linewidth=2, markersize=8, color='blue', label='Compute Only')
    plt.plot(df['Processes'], df['Total_Time'], 
             marker='s', linewidth=2, markersize=8, color='green', label='Total (with comm)')
    
    plt.title(f'Strong Scaling: Execution Time - {total_lines} items', fontsize=14)
    plt.xlabel('Number of MPI Processes', fontsize=12)
    plt.ylabel('Time (seconds)', fontsize=12)
    plt.xscale('log', base=2)
    plt.yscale('log')
    plt.xticks(df['Processes'], df['Processes'])
    plt.legend()
    plt.grid(True, which="both", ls="-", alpha=0.5)
    plt.tight_layout()
    plt.savefig(f'{CHARTS_DIR}/strong_scaling_time.png', dpi=300)
    print(f"  Saved: strong_scaling_time.png")
    plt.close()


def generate_weak_scaling_plots(csv_file):
    """
    Generate Weak Scaling charts for MPI.
    
    Weak Scaling Efficiency = T1(n) / Tp(n)
    Where:
    - T1(n) = Time for 1 process with chunk size n
    - Tp(n) = Time for P processes, each with chunk size n
    
    Since Serial_Time in CSV is for FULL problem (chunk_size * P items),
    we need to divide by P to get the base chunk time.
    """
    if not os.path.exists(csv_file):
        print(f"Warning: {csv_file} not found.")
        return
    
    df = pd.read_csv(csv_file)
    df = df.sort_values('Processes')
    total_lines = extract_lines_from_filename(csv_file)
    
    print(f"Generating weak scaling plots for {total_lines} base items...")
    
    # CORRECT Weak Scaling Efficiency calculation:
    # T1_base = Serial_Time / P (time for 1 process to handle base chunk)
    # Efficiency = T1_base / Parallel_Time
    df['T1_base'] = df['Serial_Time'] / df['Processes']
    df['Efficiency_Compute_Correct'] = df['T1_base'] / df['Compute_Time']
    df['Efficiency_Comm_Correct'] = df['T1_base'] / df['Compute_Communication_Time']
    
    # 4. Weak Scaling Efficiency (Compute Only)
    plt.figure(figsize=(10, 6))
    plt.plot(df['Processes'], df['Efficiency_Compute_Correct'], 
             marker='o', linewidth=2, markersize=8, color='blue', label='Weak Efficiency (Compute)')
    plt.axhline(y=1.0, color='red', linestyle='--', linewidth=1.5, label='Ideal Scaling (1.0)')
    
    plt.title('Weak Scaling: Efficiency (Compute Only)', fontsize=14)
    plt.xlabel('Number of MPI Processes', fontsize=12)
    plt.ylabel('Efficiency (T1 / Tp)', fontsize=12)
    plt.xscale('log', base=2)
    plt.xticks(df['Processes'], df['Processes'])
    plt.ylim(0, 1.4)
    plt.legend()
    plt.grid(True, which="both", ls="-", alpha=0.5)
    plt.tight_layout()
    plt.savefig(f'{CHARTS_DIR}/weak_scaling_efficiency_compute.png', dpi=300)
    print(f"  Saved: weak_scaling_efficiency_compute.png")
    plt.close()
    
    # 5. Weak Scaling Efficiency (With Communication)
    plt.figure(figsize=(10, 6))
    plt.plot(df['Processes'], df['Efficiency_Comm_Correct'], 
             marker='o', linewidth=2, markersize=8, color='green', label='Measured Weak Efficiency')
    plt.axhline(y=1.0, color='red', linestyle='--', linewidth=1.5, label='Ideal Scaling (1.0)')
    
    plt.title('Weak Scaling: Efficiency vs. Processes', fontsize=14)
    plt.xlabel('Number of MPI Processes', fontsize=12)
    plt.ylabel('Efficiency Value (T1 / Tp)', fontsize=12)
    plt.xscale('log', base=2)
    plt.xticks(df['Processes'], df['Processes'])
    plt.ylim(0, 1.4)
    plt.legend()
    plt.grid(True, which="both", ls="-", alpha=0.5)
    plt.tight_layout()
    plt.savefig(f'{CHARTS_DIR}/weak_scaling_efficiency.png', dpi=300)
    print(f"  Saved: weak_scaling_efficiency.png")
    plt.close()
    
    # 6. Weak Scaling Execution Times
    plt.figure(figsize=(10, 6))
    plt.plot(df['Processes'], df['Compute_Time'], 
             marker='o', linewidth=2, markersize=8, color='blue', label='Pure Compute')
    plt.plot(df['Processes'], df['Compute_Communication_Time'], 
             marker='s', linewidth=2, markersize=8, color='green', label='Total (with comm)')
    
    plt.title('Weak Scaling: Execution Times', fontsize=14)
    plt.xlabel('Number of MPI Processes', fontsize=12)
    plt.ylabel('Time (seconds)', fontsize=12)
    plt.xscale('log', base=2)
    plt.xticks(df['Processes'], df['Processes'])
    plt.legend()
    plt.grid(True, which="both", ls="-", alpha=0.5)
    plt.tight_layout()
    plt.savefig(f'{CHARTS_DIR}/weak_scaling_times.png', dpi=300)
    print(f"  Saved: weak_scaling_times.png")
    plt.close()


def generate_comparison_plot(csv_files, output_name='strong_comparison_mpi.png'):
    """
    Generate comparison plot across multiple dataset sizes.
    """
    plt.figure(figsize=(10, 6))
    
    for csv in csv_files:
        if not os.path.exists(csv):
            print(f"Skipping: {csv} (not found)")
            continue
        
        df = pd.read_csv(csv)
        df = df.sort_values('Processes')
        
        size = extract_lines_from_filename(csv)
        size_label = f"{size//1000}K" if size < 1000000 else f"{size//1000000}M"
        
        plt.plot(df['Processes'], df['Speedup_Comm'], 
                 marker='o', linewidth=2, markersize=8, label=f'N = {size_label}')
    
    # Ideal line
    max_p = df['Processes'].max()
    min_p = df['Processes'].min()
    plt.plot([min_p, max_p], [min_p, max_p], 
             'r--', linewidth=1.5, alpha=0.7, label='Ideal')
    
    plt.title('Strong Scaling Comparison: Speedup vs. Dataset Size', fontsize=14)
    plt.xlabel('Number of MPI Processes', fontsize=12)
    plt.ylabel('Speedup Factor', fontsize=12)
    plt.xscale('log', base=2)
    plt.xticks(df['Processes'], df['Processes'])
    plt.legend()
    plt.grid(True, which="both", ls="-", alpha=0.5)
    plt.tight_layout()
    plt.savefig(f'{CHARTS_DIR}/{output_name}', dpi=300)
    print(f"  Saved: {output_name}")
    plt.close()


if __name__ == "__main__":
    print("=" * 60)
    print("MPI Count Sketch - Generating Benchmark Charts")
    print("=" * 60)
    
    # Main dataset (8M = 8388608)
    perf_8m = 'results/performance_mpi_8388608.csv'
    weak_8m = 'results/weak_scaling_mpi_8388608.csv'
    
    # Additional sizes for comparison
    perf_1m = 'results/performance_mpi_1048576.csv'
    perf_500k = 'results/performance_mpi_524288.csv'
    
    # Generate main charts from 8M dataset
    print("\n--- Strong Scaling Charts ---")
    generate_strong_scaling_plots(perf_8m)
    
    print("\n--- Weak Scaling Charts ---")
    generate_weak_scaling_plots(weak_8m)
    
    # Generate comparison chart
    print("\n--- Comparison Charts ---")
    generate_comparison_plot(
        [perf_500k, perf_1m, perf_8m],
        output_name='strong_comparison_mpi.png'
    )
    
    print("\n" + "=" * 60)
    print("Chart generation complete!")
    print(f"Charts saved to: {CHARTS_DIR}/")
    print("=" * 60)
